include/unicorn/x86.h

```diff
@@ -8,6 +8,15 @@
 extern "C" {
 #endif
 
+//Memory-Management Register fields (idtr, gdtr, ldtr, tr)
+//borrow from SegmentCache in qemu/target-i386/cpu.h
+typedef struct x86_mmr {
+    uint32_t selector;
+    uint64_t base;   /* handle 32 or 64 bit CPUs */
+    uint32_t limit;
+    uint32_t flags;
+} x86_mmr;
+
```

Comment:
Internal to qemu, cpu->gdt and cpu->ldt are declared as SegmentCache (see qemu/target-i386/cpu.h line 861). Unless you want to introduce a new data type in the qemu code as well, then within the qemu code we are obligated to cast the caller's pointer to SegmentCache, so our data type needs to be identical to SegmentCache.

Using these registers is something for advanced users. The caller will need to understand what is and is not valid.

In the supplied change to qemu/target-i386/unicorn.c flags and selector are ignored when reading or writing GDTR and IDTR.

On 2/4/2016 5:05 PM, Nguyen Anh Quynh wrote:
> In include/unicorn/x86.h <https://github.com/unicorn-engine/unicorn/pull/416#discussion_r51963805>:
> 
>> @@ -8,6 +8,15 @@
>>  extern "C" {
>>  #endif
>>  
>> +//Memory-Management Register fields (idtr, gdtr, ldtr, tr)
>> +//borrow from SegmentCache in qemu/target-i386/cpu.h
>> +typedef struct x86_mmr {
>> +    uint32_t selector;
>> +    uint64_t base;   /* handle 32 or 64 bit CPUs */
>> +    uint32_t limit;
>> +    uint32_t flags;
>> +} x86_mmr;
>> +
> 
> i can see that you are trying to use the same struct for all related instructions. however, IDTR & GDTR do not have |limit| and |flags|, so this is confused. should we have separate structs for them?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/unicorn-engine/unicorn/pull/416/files#r51963805>.
>